src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php

```diff
@@ -161,46 +149,80 @@ protected function doDelete(array $ids, array $tagData = []): bool
+            }
+
+            foreach ($tagIdsByConnection as $connection) {
+                $slot = $tagIdsByConnection[$connection];
+                $connection = new \Predis\Client($connection, $this->redis->getOptions());
+                $movedTagSetIds = array_merge($movedTagSetIds, $this->renameKeys($connection, $slot->getArrayCopy()));
+            }
+        }
+
+        // No Sets found
+        if (!$movedTagSetIds) {
             return false;
         }
 
-        // Pop all tag info at once to avoid race conditions
-        $tagIdSets = $this->pipeline(static function () use ($tagIds) {
-            foreach ($tagIds as $tagId) {
-                // Client: Predis or PHP Redis 3.1.3+ (https://github.com/phpredis/phpredis/commit/d2e203a6)
-                // Server: Redis 3.2 or higher (https://redis.io/commands/spop)
-                yield 'sPop' => [$tagId, self::POP_MAX_LIMIT];
+        // Now safely take the time to read the keys in each set and collect ids we need to delete
+        $tagIdSets = $this->pipeline(static function () use ($movedTagSetIds) {
+            foreach ($movedTagSetIds as $movedTagId) {
+                yield 'sMembers' => [$movedTagId];
             }
         });
 
-        // Flatten generator result from pipeline, ignore keys (tag ids)
-        $ids = array_unique(array_merge(...iterator_to_array($tagIdSets, false)));
+        // Return combination of the temporary Tag Set ids and their values (cache ids)
+        $ids = array_merge($movedTagSetIds, ...iterator_to_array($tagIdSets, false));
 
         // Delete cache in chunks to avoid overloading the connection
-        foreach (array_chunk($ids, self::BULK_DELETE_LIMIT) as $chunkIds) {
+        foreach (array_chunk(array_unique($ids), self::BULK_DELETE_LIMIT) as $chunkIds) {
             $this->doDelete($chunkIds);
         }
 
         return true;
     }
 
-    private function redisServerSupportSPOP(): bool
+    /**
+     * Renames several keys in order to be able to operate on them without risk of race conditions.
+     *
+     * Filters out keys that do not exist before returning new keys.
+     *
+     * @see https://redis.io/commands/rename
+     *
+     * @return array Filtered list of the valid moved keys (only those that existed)
+     */
+    private function renameKeys($connection, array $ids): array
     {
-        if (null !== $this->redisServerSupportSPOP) {
-            return $this->redisServerSupportSPOP;
-        }
-
-        foreach ($this->getHosts() as $host) {
-            $info = $host->info('Server');
-            $info = isset($info['Server']) ? $info['Server'] : $info;
-            if (version_compare($info['redis_version'], '3.2', '<')) {
-                CacheItem::log($this->logger, 'Redis server needs to be version 3.2 or higher, your Redis server was detected as '.$info['redis_version']);
-
-                return $this->redisServerSupportSPOP = false;
+        // 1. Due to Predis exception we don't do this in pipeline
```

Comment:
Now that we group ids per node, we could maybe use pipeline again?
But I don't know how to deal with failures to rename to maybe not, I'll let you check this :)

Reply:
would there be any failures to rename? (I'm only aware of cross slot failure, which won't be an issue here or?)

Reply:
aha, the predis exception.

Seems like it is still valid in case of 1. non cluster predis usage and also 2. predis cluster given you don't seem to filter out tags that don't exits _(? or do you, your code is hard to read here TBH)_

Reply:
I don't filter tags that don't exist - and it's actually impossible to do, because of race conditions.
(the code is just grouping ids per node...)

Reply:
ok, well exception happens when key does not exist. So still valid.
Unsure why it throws over giving a boolean status response like phpredis does, could argue its a flaw of handling here so we should report it so it hopefully gets fixed down the line.